app/views/lead_providers/your_schools/index.html.erb

```diff
@@ -0,0 +1,11 @@
+<%= render ::SchoolInformation::YourSchools::HeadingComponent.new %>
+<%= render ::SchoolInformation::YourSchools::NavigationComponent.new(cohorts: @cohorts, selected_cohort: @selected_cohort) %>
+<%= render ::SchoolInformation::YourSchools::ParticipantsComponent.new %>
```

Comment:
I think this is the only place those are used. I am not familiar with the big picture of service, but I would not turn things into components until they are used in two places.

That said, I'm not gonna tell you to rewrite this :D

Reply:
Initially, I felt strange refactoring it into components - what I like about them is that they can be tested independently like a lego block. It has specified numbers of inputs which makes it predictable - as long as you put X inputs - you will get Y output. However, lets look at it from the opposite perspective - you are right that those are only being used once in here - they do however show the pattern with specific example. Do we want to go this road - and where it leads to - I don't know - it depends on the area and what kind of architecture we want to promote.

Reply:
Most react devs I know turn every kind of UI widget into a component and then compose them to build more complicated UIs. Another advantage I see with this approach is that  the action template, app/views/lead_providers/your_schools/index.html.erb, is a lot cleaner. Given all the effort he's invested plus these advantages I would keep the components.

Reply:
I agree we should probably keep them, but hey, I was one of those React devs not so long ago :D 

One scary thing about making something into a component straight away is that you eventually want to reuse it, but the use case doesn't quite match how it was written in the first place. So you have to refactor it to reap the benefit. Sometimes quite extensively. And I think in most cases it is easier to extract the common part from two separate places into a new component, than to refactor a component into one that matches the common logic in two places.

And another scary outcome is a complicated and interconnected forest of components that makes it really hard to understand what's going on and where.

So yeah, I think we can keep those, but we should be careful about this approach. If a page is deleted, it's easy to forget about deleting the component that is only used on that page - and we end up with a dead component that crowds the space (for example).

Reply:
Completely agree @kicferk1 . Perhaps going forward we should only introduce components when there's a need to reuse. The other reason why I was in favour of them, back in the day when I was an ember developer, was that they were more convenient for testing content than say a feature / end-to-end spec.

Reply:
@kicferk1 @amazimbe I hear your concerns. 1. Will we end up with forest of interconnected components - no if we keep them flat or alternatively namespaced. For example a rule that a component will not include another component - a view only have components might do the trick. 2. What would be a good policy to follow when building component architecture - there are scary bits , there are potential downfall - if we would follow more elegant path what would optimal component based architecture would look like. 3. I wouldnt worry too much about dead components - ts quite easy to find if anything is referencing them. Killing zombie components is trivial. 4. In terms of component testing - I am with Alfred with this - this makes testing easier and much more predictable

Reply:
Hehe, given that testing it will be its own PR, I look forward to you regretting your words that testing will be easier 😂 

As for 1-2, I think there is an implicit tradeoff here - the more we separate things out (using namespaces, or convention), the less reusability we get - and so we lose some reusability and maintainability benefits of components.  And the less separate we keep things, the more prone to interconnected forest it becomes.

The general rule of thumb I have when deciding "should something be a component?" is "will I get the bang for my buck - is it gonna be reusable, is it gonna make maintaining easier, is it something I want to encapsulate?". I am not sure the components here would hit the reusability aspects, but maybe they justify themselves in encapsulation?